File: forms/fertility-assessment/back-end-with-loco/src/engine/fertility_grader.rs
//! Fertility grader module.

use super::fertility_rules::all_rules;
use super::flagged_issues::detect_additional_flags;
use super::types::{AssessmentData, ConcernLevel, FiredRule, GradingResult};

/// Classify a numeric concern score into a NICE CG156 concern level.
///
/// Cutoffs (mirroring `fertility-grader.js`):
///   0..=2  -> low concern
///   3..=6  -> moderate concern
///   >= 7   -> high concern
pub fn classify_concern_score(score: i32) -> ConcernLevel {
    if score >= 7 {
        "high".to_string()
    } else if score >= 3 {
        "moderate".to_string()
    } else {
        "low".to_string()
    }
}

/// Evaluate every rule against the data, returning the rules that fired
/// along with the total weighted concern score.
pub fn run_rules(data: &AssessmentData) -> (i32, Vec<FiredRule>) {
    let mut total: i32 = 0;
    let mut fired: Vec<FiredRule> = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            total += rule.weight;
            fired.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                score: rule.weight,
            });
        }
    }
    (total, fired)
}

/// Pure function: grades a NICE CG156 fertility assessment.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let (concern_score, fired_rules) = run_rules(data);
    let concern_level = classify_concern_score(concern_score);
    let additional_flags = detect_additional_flags(data);

    GradingResult {
        concern_score,
        concern_level,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

File: forms/fertility-assessment/front-end-with-svelte/src/lib/engine/fertility-grader.ts
// NICE CG156 Fertility grader. Pure functions: take an `AssessmentData`
// object, fire each rule, and produce an overall concern level (Low /
// Moderate / High) plus the audit trail of fired rules.
//
// Concern level cutoffs (sum of fired rule weights):
//   0..2   -> low concern
//   3..6   -> moderate concern
//   >= 7   -> high concern

import type { AssessmentData, ConcernLevel, FiredRule, GradingResult } from './types';
import { fertilityRules } from './rules';
import { detectAdditionalFlags } from './flagged-issues';

/** Classify a numeric concern score into a category. */
export function classifyConcernScore(score: number): ConcernLevel {
	if (score >= 7) return 'high';
	if (score >= 3) return 'moderate';
	return 'low';
}

/**
 * Pure function: evaluates all NICE CG156 fertility rules against patient data.
 * Returns the concern score, concern level, fired rules, and flagged issues.
 */
export function calculateConcern(data: AssessmentData): GradingResult {
	const firedRules: FiredRule[] = [];
	let totalScore = 0;

	for (const rule of fertilityRules) {
		try {
			if (rule.evaluate(data)) {
				totalScore += rule.weight;
				firedRules.push({
					id: rule.id,
					category: rule.category,
					description: rule.description,
					score: rule.weight
				});
			}
		} catch (e) {
			// Rule evaluation failed - log for debugging but continue grading.
			console.warn(`Fertility rule ${rule.id} evaluation failed:`, e);
		}
	}

	const concernLevel = classifyConcernScore(totalScore);
	const additionalFlags = detectAdditionalFlags(data);

	return {
		concernScore: totalScore,
		concernLevel,
		firedRules,
		additionalFlags,
		timestamp: new Date().toISOString()
	};
}
